### ISSUE TITLE
Linter that (private) typedefs can be removed in favor of the new function syntax

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#884]
The Function syntax introduced in Dart 1.24 can be used to replace many typedefs that exist only to spell out types in internal interfaces.

Right now, the only signal that a typedef could be replaced with an inline Function is that the typedef is _private 🙁  but that's a start! If the linter has package-level knowledge, it could provide the same rule for typedefs that are not exported into the "public API."

**Comment by a14n** [comment]
I started to create this lint but after trying it on Flutter I think we should restrict it to lint private typedefs that are used only once.

```dart
typedef void _F(int i, double d); // OK
_m1(_F f);
_m2(_F f);
_m3(_F f);

typedef void _F2(); // LINT because only one usage
_mA(_F2 f);
```

WDYT ?

**Comment by srawlins** [comment]
Ah, makes sense. I like it. Probably still many existing occurrences.

**Comment by a14n** [comment]
>they mean you don't have to duplicate potentially complicated signatures in multiple places, so making a change in one place fixes all the places that use that signature

The lint only occurs on _private typedefs that are only used once_.

So once you face a duplicated function signature with the same purpose you are allowed to create a typedef (in that the lint will not bother you).

>they document the meaning of the signature, both by having a name, and by giving you somewhere to provide documentation about the contract of the particular signature.

For _private typedefs that are only used once_ the documentation can live on the place where it is used.

```dart
/// A callback called when a click occurs.
typedef _OnClickCallback = void Function();

/// Register an [_OnClickCallback] that is called on every click.
void _register(_OnClickCallback onClick) {...}



/// Register a function that is called on every click.
void _registerWithoutTypedef(void Function() onClick) {...}
```

As a user of those methods the second version is shorter and seems easier to reason about because you don't have to jump over the source to fully understand what is needed.

**Comment by Hixie** [comment]
I've no objection to the lint existing, I just don't think it's a particularly useful one for Flutter.
